pyHS100/antitheft.py

```diff
@@ -0,0 +1,199 @@
+        "year": int(0), 0 for all years
+        "longitude": int/float(0), 0 for all locations
+        "day": int(0), 0 for all days of the month
+        "force": int(0), 0 or 1, 1 to force schedule
+        "latitude": int/float(0), 0 for all locations
+        "emin": int(0)
+
+        :param str name: Name of the rule
+        :param dict rule: Parameters for the rule (see description)
+        :return: None
+        :rtype: None
+        :raises SmartPlugException: on error
+        """
+
+        if name in ('_device', '_query_helper'):
+            object.__setattr__(self, name, rule)
+            return
+
+        rule['name'] = name
+
+        if name in self:
+            rule['id'] = self[name]['id']
+            self._query_helper(
+                "anti_theft",
+                "edit_rule",
+                rule
+            )
+        else:
+            self._query_helper(
+                "anti_theft",
+                "add_rule",
+                rule
+            )
+
+    def __delitem__(self, name):
+        """
+        Deletes rule.
+
+        :param str name: Name of the rule
+        :return: None
+        :rtype: None
+        :raises NoAntiTheftFound: If rule not found
+        :raises SmartPlugException: on error
+        """
+
+        for n, a in self:
+            if n == name:
+                return self._query_helper(
+                    "anti_theft",
+                    "delete_rule",
+                    {"id": a['id']}
+                )
+        raise NoAntiTheftFound(name)
+
+    def get(self, name=None, *args):
+        """
+        Gets rule
+
+        :param str name: Name of the rule.
+        :param None name: If None returns all rules.
+        :param args: Used for default return value.
+        :return: dict name: All rules if item is None else rule
+        :rtype: dict
+        :raises NoAntiTheftFound: If rule not found
+        :raises SmartPlugException: on error
+        """
+
+        if name is None:
+            return dict(list(antitheft for antitheft in self))
+        elif args and name not in self:
+            return args[0]
+        else:
+            return self[name]
+
+    def __getitem__(self, name):
+        """
+        Gets rule
+
+        :param str name: Name of the rule.
+        :return: dict rule
+        :rtype: dict
+        :raises NoAntiTheftFound: If rule not found
+        :raises SmartPlugException: on error
+        """
+
+        if name in self.__dict__:
+            return self.__dict__[name]
+
+        for n, a in self:
+            if n == name:
+                return a
+
+        raise NoAntiTheftFound(name)
+
+    def __del__(self):
+        """
+        Deletes all rules.
+
+        :return: None
+        :rtype: None
+        :raises SmartPlugException: on error
+        """
+
+        self._query_helper(
+            "anti_theft",
+            "delete_all_rules"
+        )
+        self._device.new_antitheft()
+
+
+class NoAntiTheftFound(Exception):
```

Comment:
that i would have done but because SmartPlugException class is in pyHS100.py which in the same file where the 3 new classes are imported into so there was now way i would have been able to import it into any of those files to create a subclass of it due to circular imports. and i couldn't have created the subclasses in the pyHS100.py file either due to the same thing. so unless there is a separate exception file created with SmartPlugException in it then i am not sure how to go about subclassing it.